modules/prompt_parser.py

```diff
@@ -96,48 +97,324 @@ def get_schedule(prompt):
     return [promptdict[prompt] for prompt in prompts]
 
 
-ScheduledPromptConditioning = namedtuple("ScheduledPromptConditioning", ["end_at_step", "cond"])
-ScheduledPromptBatch = namedtuple("ScheduledPromptBatch", ["shape", "schedules"])
+re_compound_tokenizer = re.compile(r'''
+\\\\|\\(|\\)|(?P<e>\(|\)|\bPLUS\b|\bAND\b|\bNOT\b)|
+:\s*(?P<w>[+-]?(?:\d+\.\d*|\.\d+|\d+))
+|\w+|.
+''', re.VERBOSE)
 
+def parse_compound(prompt):
+    """
+    Parse a prompt that may include AND, NOT and :weight arguments
+    into a list of lists of weighted subprompts that will be evaluated
+    for it.
+
+    Used to implement composition, as described in https://arxiv.org/abs/2206.01714
+    Compositional Visual Generation with Composable Diffusion Models by Liu et al.
+
+    >>> p = parse_compound
+    >>> p('prompt (no compound:.5)')
+    [[[('prompt (no compound:.5)', 1.0)], 1.0]]
+    >>> p('a AND b')
+    [[[('a', 1.0)], 0.5], [[('b', 1.0)], 0.5]]
+    >>> p('a:3 AND (b:2) :+1.4')
+    [[[('a', 1.0)], 1.5], [[('(b:2)', 1.0)], 0.7]]
+    >>> p('pre (a AND b  ) aft')
+    [[[('pre a aft', 1.0)], 0.5], [[('pre b aft', 1.0)], 0.5]]
+    >>> p('a (NOT b) c')
+    [[[('a b c', 1.0)], -0.5]]
+    >>> p('a ( NOT b) c')
+    [[[('a  c', 1.0)], 1.0], [[('a b c', 1.0)], -0.5]]
+    >>> p('a:3:2 PLUS b:3')
+    [[[('a:3', 0.4), ('b', 0.6)], 1.0]]
+    >>> p('a:3 PLUS b:4 AND c:4 PLUS d:6 :10')
+    [[[('a', 0.75), ('b', 0.25)], 2.0], [[('c', 0.4), ('d', 0.6)], 5.0]]
+    >>> p('a (b AND c) e (f AND g)')
+    [[[('a b e f', 1.0)], 0.5], [[('a c e g', 1.0)], 0.5]]
+    >>> p('a (b AND c) (NOT d)')
+    [[[('a b d', 1.0)], -0.5], [[('a c ', 1.0)], 1.0]]
+    >>> p('a (b AND c) ( NOT d)')
+    [[[('a b ', 1.0)], 1.0], [[('a c d', 1.0)], -0.5]]
+    >>> p('a (b AND c) (d NOT e NOT f)')
+    [[[('a b d', 1.0)], 1.0], [[('a c e', 1.0)], -0.25], [[('a b f', 1.0)], -0.25]]
+    >>> p('(a:4 AND b:3)') # degenerate cases
+    [[[('a', 1.0)], 2.0], [[('b', 1.0)], 1.5]]
+    >>> p('(a:4 AND b:3)(NOT a)')
+    [[[('aa', 1.0)], -2.0], [[('b', 1.0)], 3.0]]
+    >>> p('(a:4 AND b:3)(NOT a:6)')
+    [[[('aa', 1.0)], -2.0], [[('b', 1.0)], 3.0]]
+    >>> p('a AND (b AND c)') # ignore recursion
+    [[[('a', 1.0)], 0.5], [[('(b AND c)', 1.0)], 0.5]]
+    """
 
-def get_learned_conditioning(model, prompts, steps):
-    res = []
+    toks = []
+    last_normal = False
+    for m in re_compound_tokenizer.finditer(prompt):
+        t = m.group()
+        v = None

```

Comment:
```suggestion
        v = 1.0
```
Is there any reason to create a tuple element like `Union[float, None]` if it could just simply be `float` with default value of `1.0`? IMO it would simplify the type system / cognitive overhead. You can then also remove the `try/except` and `float` casting in `set_weights`.